tensorflow/compiler/xla/service/slice_sinker.cc

```diff
@@ -0,0 +1,276 @@
+/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/xla/service/slice_sinker.h"
+#include <algorithm>
+#include <utility>
+#include <vector>
+#include "absl/types/optional.h"
+#include "absl/container/flat_hash_set.h"
+#include "tensorflow/compiler/xla/shape_util.h"
+
+namespace xla {
+
+namespace {
+
+class SliceSinkerImpl {
+ public:
+  // Returns whether the instruction has been visited. The visited instruction
+  // has been grouped in a true user group with its peers and it should not be
+  // grouped in another true user group.
+  bool IsVisited(const HloInstruction* instruction) const;
```

Comment:
The code where instructions are added to visited_ doesn't seem to match the description of routine IsVisited here. I see in some situations, we add an instruction to visited_ when we can determine that the instruction can be  transformed by this pass.
Also, if an instruction is an ElementwiseOperationOnSlices, and we add it to a group and transform it. Then after the transformation, the instruction will have zero users.  When I think about that, I believe we don't even need member variable visited_ for the purpose you describe here.

Reply:
As you said, visited_ is not indispensible. visited_ is just to avoid redundant match.

Some ElementwiseOperationOnSlices will be transformed, but they are still in graph, util removed by DCE. They will still be visited to check whether in another group. Additionally, other ElementwiseOperationOnSlices will be filtered by cost function ShouldReplace.

Reply:
The vector visited_ is not identifier whether the instruction could be transformed. The visited_ is the identifier whether the instruction has been transformed or could never be transformed. If the instruction has transformed or could never be transformed, it need not to be consider transforming again.

Reply:
Yes, I understand this. An instruction that is already transformed has a user_count == 0, you can use this fact to avoiding transforming it again.  If during the call to FindPeerElementwiseOperations for instruction `A`, we find out instruction `B`,  which is after `A` in the post order, is not transformable, we currently add `B` to visited_. This is the situation where using visited_ can avoid inspecting `B` again.  On the other hand, if we don't add `B` to visited_,  either FindElementwiseOperationOnSlices or FindPeerElementwiseOperations for `B` will return nullopt. So the question is whether we want to introduce a hash that will store all the instructions at the end of this process to save the time of making more calls to FindElementwiseOperationOnSlices or FindPeerElementwiseOperations. If we assume elemementwise-operation-with-all-slices-operands are very rare in a normal program, FindElementwiseOperationOnSlices should quickly return nullopt, and we should go for the approach of not using the visited_ hash. What do you think?